File: cracking_the_coding_interview/a_arrays_and_strings/e/_1_5.py
import os
from dotenv import load_dotenv
import json


def one_away(first, second):
    shorter_string = first if len(first) <= len(second) else second
    longer_string = second if len(first) <= len(second) else first
    if len(longer_string) - len(shorter_string) > 1:
        return False
    index_shorter = 0
    index_longer = 0
    difference_found = False

    while index_shorter < len(shorter_string):
        if shorter_string[index_shorter] != longer_string[index_longer]:
            if difference_found:
                return False
            else:
                difference_found = True
            if len(shorter_string) != len(longer_string):
                index_longer += 1
            else:
                index_longer += 1
                index_shorter += 1
        else:
            index_longer += 1
            index_shorter += 1

    return True

"""
FAIL:
This solution does not account for the order of the characters that need to match as well

def one_away(first, second):
    tracker = {}
    if abs(len(first) - len(second)) > 1:
        return False
    for letter in first:
        if letter in tracker:
            tracker[letter] += 1
        else:
            tracker[letter] = 1
    for letter in second:
        if letter in tracker:
            tracker[letter] -= 1
        else:
            tracker[letter] = 1
        if tracker[letter] == 0:
            del tracker[letter]
    if len(tracker) == 0:
        return True
    elif len(tracker) <= 2 and all(value == 1 for value in tracker.values()):
        return True
    else:
        return False
"""

if __name__ == "__main__":
    print(one_away("asdfasdb", "asdfasdac"))

    # TEST SCRIPT
    load_dotenv()

    CURRENT_DIRECTORY = os.getcwd()
    dir_to_book = r"\cracking_the_coding_interview"
    dir_to_chapter = r"\a_arrays_and_strings"
    dir_to_problem = r"\e"
    file_name = r"\test.json"

    dir_to_test_file = (
        rf"{CURRENT_DIRECTORY}{dir_to_book}{dir_to_chapter}{dir_to_problem}{file_name}"
    )

    f = open(dir_to_test_file)
    data = json.load(f)

    for input in data:
        print(f"#{input["index"]}: {one_away(input["input"]["first"], input["input"]["second"])== input["output"]}")

File: cracking_the_coding_interview/a_arrays_and_strings/e/_1_5.js
// prettier-ignore
import test from "./test.json" with {type: "json"};

const one_away = (first, second) => {
  const shorter_string = first.length >= second.length ? second : first;
  const longer_string = first.length >= second.length ? first : second;
  let is_one_away = true;
  if (longer_string.length - shorter_string.length > 1) {
    return false;
  }
  let index_shorter = 0;
  let index_longer = 0;
  let difference_found = false;

  while (index_shorter < shorter_string.length) {
    if (shorter_string[index_shorter] !== longer_string[index_longer]) {
      if (difference_found) {
        is_one_away = false;
        break;
      } else {
        difference_found = true;
      }
      if (shorter_string.length !== longer_string.length) {
        index_longer++;
      } else {
        index_longer++;
        index_shorter++;
      }
    } else {
      index_longer++;
      index_shorter++;
    }
  }
  return is_one_away;
};

//#1
const solution = one_away;
if (true) {
  let errors = 0;
  let error_indices = [];
  test.forEach((inputLine) => {
    const input = inputLine["input"];
    //#2
    const output = solution(input["first"], input["second"]);
    if (!output === inputLine["output"]) {
      console.log(`There is an error at #${inputLine["index"]}`);
      console.log(`My answer was: ${output}`);
      console.log(`Given ans was: ${inputLine["output"]} `);
      errors++;
      error_indices.push(inputLine["index"]);
    }
  });
  if (errors) {
    console.log(`There were ${errors} errors`);
    error_indices.forEach((error) => {
      console.log(`Please check #${error}`);
    });
  } else {
    console.log("No errors were found");
  }
} else {
  //#3
  const index = 13;
  const inputLine = test.find((inputLine) => inputLine.index === index);
  const input = inputLine["input"];
  Object.keys(input).forEach((key) => {
    console.log(`The argument for "${key}" is:           ${input[key]}`);
  });
  //#4
  const outcome = solution(input["first"], input["second"]);
  console.log("END:", outcome);
}
